lib/Feature/FeatureModelBuilder.cpp

```diff
@@ -0,0 +1,89 @@
+//                        FeatureModelBuilder
+//===----------------------------------------------------------------------===//
+
+/// Decide whether two features are mutual exclusive. Beware that this method
+/// only detects very simple trees with binary excludes.
+bool isSimpleMutex(const Feature *A, const Feature *B) {
+  return std::any_of(
+      A->excludes().begin(), A->excludes().end(), [A, B](const auto *E) {
+        if (const auto *LHS =
+                llvm::dyn_cast<PrimaryFeatureConstraint>(E->getLeftOperand());
+            LHS) {
+          if (const auto *RHS = llvm::dyn_cast<PrimaryFeatureConstraint>(
+                  E->getRightOperand());
+              RHS) {
+            // A excludes B
+            if (LHS->getFeature() &&
+                LHS->getFeature()->getName() == A->getName() &&
+                RHS->getFeature() &&
+                RHS->getFeature()->getName() == B->getName()) {
+              return std::any_of(
+                  B->excludes().begin(), B->excludes().end(),
+                  [A, B](const auto *E) {
+                    if (const auto *LHS =
+                            llvm::dyn_cast<PrimaryFeatureConstraint>(
+                                E->getLeftOperand());
+                        LHS) {
+                      if (const auto *RHS =
+                              llvm::dyn_cast<PrimaryFeatureConstraint>(
+                                  E->getRightOperand());
+                          RHS) {
+                        // B excludes A
+                        return LHS->getFeature() &&
+                               LHS->getFeature()->getName() == B->getName() &&
+                               RHS->getFeature() &&
+                               RHS->getFeature()->getName() == A->getName();
+                      }
+                    }
+                    return false;
+                  });
+            }
+          }
+        }
+        return false;
+      });
+}
+
+bool FeatureModelBuilder::detectXMLAlternatives() {
+  for (auto *F : FM->features()) {
+    auto Children = F->getChildren<Feature>();
+    if (Children.size() > 1 &&
+        std::all_of(Children.begin(), Children.end(), [Children](auto *F) {
+          for (auto *C : Children) {
+            if (F == C) {
+              continue;
+            }
+            if (!isSimpleMutex(F, C)) {
+              return false;
+            }
+          }
+          return !F->isOptional();
+        })) {
+      Special.addRelationship(Relationship::RelationshipKind::RK_ALTERNATIVE,
+                              F);
+    }
+  }
+  return true;
+}
+
+std::unique_ptr<FeatureModel> FeatureModelBuilder::buildFeatureModel() {
+  if (!FM->getRoot()) {
+    auto T = FeatureModelModifyTransaction::openTransaction(*FM);
+    T.setRoot(std::make_unique<RootFeature>("root"));
+    if (!T.commit()) {
+      return nullptr;
+    }
+  }
+  return Features.commit() && Transactions.commit() &&
```

Comment:
What is the meaning of Transactions here?

Reply:
These are dependent stages, if we want to assure, all feature reference can be resolved.

1. `Features` are all emplacements of feature nodes into the FM, without any structural information yet -- basically everything is child of root.
2. `Transactions` contains all modification to build the tree structure of features, like parent / child relations.
3. `PostTransactions` (we may use a better name, see above) insert relationships an constraints. At this stage we can assume, that all feature nodes and edges are already present, so we can easily change the hierarchy or let constraints refer to existing features (through this could also be done in the prior stage).
4. `Special` currently only handles XML alternative groups, so we could move this locally, but I think providing a collection for similar corner cases might be useful.